qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+from qiskit.circuit.library.generalized_gates.pauli import PauliGate
+from qiskit.circuit.library.standard_gates.p import PhaseGate
+from qiskit.circuit.library.standard_gates.rx import RXGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.library.standard_gates.u1 import U1Gate
+from qiskit.circuit.quantumregister import QuantumRegister, Qubit
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit, DAGInNode, DAGNode, DAGOutNode
+from qiskit.quantum_info import Pauli
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.optimization.commutation_analysis import \
+    CommutationAnalysis
+from qiskit.transpiler.passmanager import PassManager
+
+commutator = CommutationChecker(standard_gates_commutations)
+
+
+class LightCone(TransformationPass):
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can
+    handle either a Pauli operator one would like to measure or a measurement on a set
+    of qubits.
+    """
+
+    def __init__(self, observable: Pauli | None = None):
+        """
+        LightCone initializer.
+
+        Args:
+            observable: If None the lightcone will be computed for the set
+            of measurements in the circuit.
+            If a Pauli operator is specified, the lightcone will correspond to
+            the reduced circuit with the
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
+        """For now this method assumes that all the measurements in the circuit are
+        final measurements."""
+        qubits_measured = set()
+        for node in dag.topological_nodes():
+            # print(getattr(node, "name", False))
+            if getattr(node, "name", False) == "measure":
+                qubits_measured |= {x for x in node.qargs}
+        print(qubits_measured)
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[int], tuple[Instruction, list[int]]]:
+        """Returns the initial lightcone.
+        If obervable is None, the lightcone is a different Z operator in each one of the
+        qubits with a measurement. If a Pauli observable is defined, that will be
+        the first gate in the circuit.
+        """
+        if self.observable is None:
+            qubits_measured = self._find_measurement_qubits(dag)
+            light_cone = qubits_measured
+            light_cone_ops = [(PauliGate("Z"), [qubit_index]) for qubit_index in qubits_measured]
+        else:
+            pauli_string = str(self.observable)
+            light_cone = [dag.qubits[i] for i, p in enumerate(pauli_string) if p != "I"]
```

Comment:
The string representation of a Pauli in Qiskit puts the qubit-0 term at the right end of the string*. So I believe this list comprehension incorrectly reverses the observable and must be fixed.

Example of the problem here. Observable has `Z` on qubit 1 (bottom qubit in drawing), which does NOT commute with the CX gate, but the indexing problem puts `Z` on the wrong qubit so the transpiler pass concludes that it DOES commute with the CX gate:
<img width="512" alt="image" src="https://github.com/user-attachments/assets/f84d5223-1b9a-4214-b632-aaaeb190dec4">
(I suppressed a debug print statement in the code)

Since this kind of bug is extremely common when working with `str` representations of Pauli strings, in general I **strongly** suggest not working with the `str` representation of Paulis unless/until you absolutely have to. A safer approach is to index the `Pauli` object directly: 
```
pauli = Pauli('XYZ')
bit_idx = 2
pauli[bit_idx] ## this returns Pauli('X'), which is correct
str(pauli)[bit_idx] ## this returns Pauli('Z'), which is incorrect
```
Indexing the Pauli directly interfaces with the numpy arrays backing the Pauli object, so in addition to being safer it's also often faster than working with the string representation.

\* - Often this is called being "little-endian", though I'm told that technically that term is reserved for byte-ordering, not bit-ordering, so "MSb0" or "MSbF" might be more appropriate, not sure.